libsplinter/src/circuit/template/yaml_parser/mod.rs

```diff
@@ -0,0 +1,134 @@
+    let template = Template::deserialize(&data, &version_guard.version)?;
+    Ok(template)
+}
+
+#[derive(Deserialize, Debug)]
+pub enum Template {
+    V0_1(v0_1::YamlCircuitCreateTemplate),
+}
+
+impl Template {
+    fn deserialize(data: &[u8], version: &str) -> Result<Self, Error> {
+        match version {
+            "v0.1" => {
+                let template = serde_yaml::from_slice::<v0_1::YamlCircuitCreateTemplate>(data)?;
+                Ok(Template::V0_1(template))
+            }
+            _ => Err(Error::new(&format!(
+                "Invalid template version: {}",
+                version
+            ))),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::fs::{remove_file, File};
+    use std::io::Write;
+    use std::{env, panic, thread};
+
+    static EXAMPLE_TEMPLATE_YAML: &[u8; 168] = br##"version: v0.1
+args:
+    - name: admin-keys
+      required: false
+      default: $(a:SIGNER_PUB_KEY)
+rules:
+    set-management-type:
+        management-type: "gameroom" "##;
+
+    /*
+     * Verifies load_template correctly loads a template version 0.1
+     */
+    #[test]
+    fn test_parse_template_v0_1() {
+        run_test(|test_yaml_file_path| {
+            write_yaml_file(test_yaml_file_path);
+            let template_version =
+                load_template(test_yaml_file_path).expect("failed to load template");
+            match template_version {
+                Template::V0_1(template) => {
+                    assert_eq!(&template.version(), "v0.1");
+                    let args = template.args();
+                    for arg in args {
+                        assert_eq!(&arg.name(), "admin-keys");
+                        assert_eq!(arg.required(), false);
+                        assert_eq!(arg.default_value(), Some("$(a:SIGNER_PUB_KEY)".into()));
+                    }
+
+                    let management_type = template
+                        .rules()
+                        .set_management_type()
+                        .expect("Management type was not deserialize correctly");
+                    assert_eq!(&management_type.management_type(), "gameroom");
+                }
+            }
+        })
+    }
+
+    fn run_test<T>(test: T) -> ()
+    where
+        T: FnOnce(&str) -> () + panic::UnwindSafe,
+    {
+        let test_yaml_file = temp_yaml_file_path();
+
+        let test_path = test_yaml_file.clone();
+        let result = panic::catch_unwind(move || test(&test_path));
+
+        remove_file(test_yaml_file).unwrap();
+
+        assert!(result.is_ok())
+    }
+
+    fn temp_yaml_file_path() -> String {
+        let mut temp_dir = env::temp_dir();
+
+        let thread_id = thread::current().id();
+        temp_dir.push(format!("test_parse_template-{:?}.yaml", thread_id));
+        temp_dir.to_str().unwrap().to_string()
+    }
+
+    fn write_yaml_file(file_path: &str) {
+        println!("file_path {}", file_path);
```

Comment:
Not sure this println is useful, since cargo doesn't print to stdout in tests by default

Reply:
cargo prints to stdout if a test fails. But yeah, in this case this was just debugging crud, so removed.